File: goblinframework-core/src/main/java/org/goblinframework/core/util/json/ObjectIdDeserializer.java
package org.goblinframework.core.util.json;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer;
import org.bson.types.ObjectId;

import java.io.IOException;

public class ObjectIdDeserializer extends StdScalarDeserializer<ObjectId> {

  public ObjectIdDeserializer() {
    super(ObjectId.class);
  }

  @Override
  public ObjectId deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
    String value = p.getValueAsString();
    if (!ObjectId.isValid(value)) {
      return null;
    }
    return new ObjectId(value);
  }
}

File: goblinframework-core/src/main/java/org/goblinframework/core/mapper/deserializer/ObjectIdDeserializer.kt
package org.goblinframework.core.mapper.deserializer

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer
import org.bson.types.ObjectId

class ObjectIdDeserializer : StdScalarDeserializer<ObjectId>(ObjectId::class.java) {

  override fun deserialize(p: JsonParser, ctx: DeserializationContext): ObjectId? {
    val value = p.valueAsString
    return if (!ObjectId.isValid(value)) null else ObjectId(value)
  }
}
